test/unittest/unittest_nntrainer_tensor_pool.cpp

```diff
@@ -0,0 +1,364 @@
+  EXPECT_NE(t2, nullptr);
+  EXPECT_FALSE(t2->isAllocated());
+
+  EXPECT_EQ(t1, t2);
+}
+
+/**
+ * @brief set batch
+ */
+TEST(TensorPool, set_batch_01_p) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::ZERO_LIFESPAN, "abc"));
+  EXPECT_NE(t1, nullptr);
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_EQ(t1->batch(), 1);
+  EXPECT_NO_THROW(pool.setBatchSize("abc", 10));
+  EXPECT_EQ(t1->batch(), 10);
+}
+
+/**
+ * @brief set batch
+ */
+TEST(TensorPool, set_batch_02_n) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::ZERO_LIFESPAN, "abc"));
+  EXPECT_NE(t1, nullptr);
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_THROW(pool.setBatchSize("not_exist", 10), std::invalid_argument);
+  EXPECT_EQ(t1->batch(), 1);
+}
+
+/**
+ * @brief zero size pool as no usage
+ */
+TEST(TensorPool, finalize_01_p) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1, *t2;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::ZERO_LIFESPAN, "abc1"));
+  EXPECT_NE(t1, nullptr);
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_NO_THROW(
+    t2 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc2"));
+  EXPECT_NE(t2, nullptr);
+  EXPECT_FALSE(t2->isAllocated());
+
+  EXPECT_NE(t1, t2);
+
+  EXPECT_NO_THROW(pool.finalize(nntrainer::BasicPlanner(), 0, 2));
+  EXPECT_EQ(pool.minMemoryRequirement(), 0);
+
+  EXPECT_FALSE(t1->isAllocated());
+  EXPECT_FALSE(t2->isAllocated());
+}
+
+/**
+ * @brief max lifespan tensors
+ */
+TEST(TensorPool, finalize_02_p) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1, *t2;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {0},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc1"));
+  EXPECT_NE(t1, nullptr);
+
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_NO_THROW(
+    t2 = pool.requestTensor(nntrainer::TensorDim({1}), {1},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc2"));
+  EXPECT_NE(t2, nullptr);
+
+  EXPECT_FALSE(t2->isAllocated());
+
+  EXPECT_NE(t1, t2);
+
+  EXPECT_NO_THROW(pool.finalize(nntrainer::BasicPlanner(), 0, 2));
+  EXPECT_EQ(pool.minMemoryRequirement(), t1->bytes() + t2->bytes());
+
+  EXPECT_FALSE(t1->isAllocated());
+  EXPECT_FALSE(t2->isAllocated());
+}
+
+/**
+ * @brief max lifespan tensors
+ */
+TEST(TensorPool, finalize_03_p) {
+  nntrainer::TensorPool pool;
+
+  EXPECT_NO_THROW(pool.finalize(nntrainer::BasicPlanner(), 0, 2));
+}
+
+/**
+ * @brief allocate
+ */
+TEST(TensorPool, allocate_deallocate_01_p) {
```

Comment:
how about using something like reserve, evict?

Reply:
for the name?

Reply:
not for the name but the concept itself... it's because this sounds like memory allocation. but after the allocate call, tensor->isAllocate() is false.

Reply:
Other than allocate will be good I think?

Reply:
> it's because this sounds like memory allocation. but after the allocate call, tensor->isAllocate() is false.

After the allocate call, tensor->isAllocate() is true. It sounds like memory allocation because it is. Let me know if I am missing what you are trying to convey here.

Reply:
I see, I think I missed the bottom part :) Thanks for the explanation!